lib/tests/test_local_working_copy.rs

```diff
@@ -546,6 +546,55 @@ fn test_checkout_discard() {
     assert!(!reloaded_wc.file_states().unwrap().contains_path(file2_path));
 }
 
+#[test]
+fn test_materialize_conflicted_files() {
```

Comment:
If I understand correctly, here you have a conflict in a single file of the form "A - B + C - B + A", and you want to simplify it to "A - B + C". I don't think we should touch this case in this PR, for two reasons:

- If we wanted to simplify such cases, I think it should be simplified on the level of the tree conflict, not file-by-file.
- I am not sure whether we want to simplify it at all.

We currently simplify "A - B + A - B + A" conflicts to "A" in 

https://github.com/martinvonz/jj/blob/fa6f5e3880b369f6d3e82457875283d8b345cac4/lib/src/merge.rs#L91-L105

but we do not simplify "A - B + C - B + A" there. There's an argument to be made that we should simplify that, but I'm not sure, and it should go there if we would simplify it.

Reply:
Another way to ask my question: is there an important use-case where such conflicts appear and need to be simplified? I may have missed one.

Reply:
> it should go there if we would simplify it

I was actually thinking that this new `deduplicate()` could be a (small) step towards making `simplify()` not have that special case. As you probably know, removing the special case from `simplify()` would mean that we can remove this hack: https://github.com/martinvonz/jj/blob/fa6f5e3880b369f6d3e82457875283d8b345cac4/cli/src/commands/squash.rs#L272-L275. But that's a large UX change that seems like it would be mostly negative (leaving "deduplicatable" conflicts in commits and making the user resolve them somehow).

> Another way to ask my question: is there an important use-case where such conflicts appear and need to be simplified? I may have missed one.

I think it's somewhat common to have the same changes from two related commits. Sometimes they end up in a conflict with some other change. I feel like I've seen that in practice several times.

Reply:
> As you probably know, removing the special case from simplify() would mean that we can remove this hack

I didn't know that; could you explain it further?

At a glance, I'm not sure why you consider that a hack. In my mind, the mental model is that `jj squash --from a --to descendant_of_a -i` should be the same as `jj split -r a`, then `jj rebase -r half_of_a --after descendant_of_a`, then `jj squash -r half_of_a` [^1]. The rebasing mentioned in the comment seems to fit very naturally in this model; it would be a natural part of `jj rebase -r half_of_a --after descendant_of_a`.

[^1]: the last squash can be thought of as being very close to a `jj restore --from half_of_a --to descendant_of_a; jj abandon half_of_a`.


> But that's a large UX change that seems like it would be mostly negative (leaving "deduplicatable" conflicts in commits and making the user resolve them somehow).

What are you referring to by "that": removing the special case? Removing `deduplicate` from this PR?

> I think it's somewhat common to have the same changes from two related commits. Sometimes they end up in a conflict with some other change. I feel like I've seen that in practice several times.

I don't have a good sense for this.

---------------------

The way I think of this is that if we want to simplify "A - B + C - B + A", we should do that everywhere, not just per-file while materializing. It seems to me like a completely separate issue. What do you think?

Reply:
BTW, I haven't read the Discord discussion thoroughly (yet?), apologies if I'm asking things that are already there.

Reply:
> > As you probably know, removing the special case from simplify() would mean that we can remove this hack
> 
> I didn't know that; could you explain it further?
> 
> At a glance, I'm not sure why you consider that a hack. In my mind, the mental model is that `jj squash --from a --to descendant_of_a -i` should be the same as `jj split -r a`, then `jj rebase -r half_of_a --after descendant_of_a`, then `jj squash -r half_of_a` [1](#user-content-fn-1-c3933d0a44de75048e2110f5afdb0987). The rebasing mentioned in the comment seems to fit very naturally in this model; it would be a natural part of `jj rebase -r half_of_a --after descendant_of_a`.

Yes, that's correct. I mean that very generally (applies to both `jj rebase` and that `jj squash` code, etc.) it's ugly that changes can disappear when you rebase them onto a commit that already has them. So if you then decide to rebase them back, the result will not be what you had before the initial rebase. If we had instead kept the conflict, then no information would be lost.

Reply:
However, it's also often very convenient that that loss happens instead of the conflict. That's the whole reason we haven't made the change. And it's presumably why Git and Mercurial also did that.

Reply:
> BTW, I haven't read the Discord discussion thoroughly (yet?), apologies if I'm asking things that are already there.

IMO, you shouldn't have to :)

Reply:
> it's ugly that changes can disappear when you rebase them onto a commit that already has them. So if you then decide to rebase them back, the result will not be what you had before the initial rebase. If we had instead kept the conflict, then no information would be lost.

+1. The words I usually use for this are "resolving `A - B  + A` conflicts to `A` breaks patch conflict theory", but your explanation might be clearer.

> However, it's also often very convenient that that loss happens instead of the conflict

+1. @bnjmnt4n , if you don't have the context, we had discussed this contradiction a few times. I don't think there's a perfect solution.

------------

I'm quite interested in @martinvonz 's opinion that `deduplicate` would help this situation. I don't follow his thinking yet.

At the same time, I think this should be a separate discussion from this PR, not in the sense that it's bad to have the discussion here, but in the sense that I'd prefer this PR did more basic resolution of per-file conflicts and didn't address the `deduplicate` question (especially if it's not too hard to split the concerns).

Reply:
> I'm quite interested in @martinvonz 's opinion that `deduplicate` would help this situation. I don't follow his thinking yet.

If we make `simplify()` not have the special case for when all sides made the same change, then we would want to have another way of doing that at a later point. This new `deduplicate()` method seems like a good fit. That's all I meant :)

Reply:
The primary reason I added this was actually because of what I found unexpected when working on the example in the tests.

The tests contain a 3-sided conflict, but the 2 files effectively only have 2 sides due to duplicated sides:
1. A - B + A - B + C
2. A - B + C - B + C

The file-level merging of (1) actually already merged A - B + A into A, so the conflict is effectively (A - B + C) but the other file still contains a 3-sided conflict, so there are 5 trees being stored. Hence, the conflicted files being stored are:
1. A - A + A - B + C (as far as I understand, do correct me if I'm wrong)
2. A - B + C - B + C

 The first file was simplified using `.simplify()`, but the second file was not (which is expected).

> Another way to ask my question: is there an important use-case where such conflicts appear and need to be simplified? I may have missed one.

Strictly no, but as an end-user, I personally found it confusing when it seems I have to repeat the merge, comparing to a base and side that I have already seen.

Reply:
You mean on of the other tests in this PR? I'll go look for it.

What do you think about splitting the concerns: just do simplifying per-file in a commit or in a PR, and then thinking about de-duplicating separately?

Reply:
Sure, I can split deduplication into another PR if that's preferred.

Reply:
It sounds like @ilyagr somewhat strongly prefers that. I would also prefer it, but I didn't ask for it because it also seems pretty harmless to me (but there may be surprises we haven't thought of) :)

@ilyagr, do you prefer a separate PR or just separate commits at the end of this PR? If it's a separate PR, what's the reason for that? Would we wait until early in the next release cycle?

Reply:
> it also seems pretty harmless to me 

It does not feel harmless to me, though I haven't yet come up with an example where deduplication makes things clearly worse (I haven't done much new thinking since my last message). Perhaps if I saw the reason you think it's harmless, I'd feel differently. I don't think it's a *terrible* idea; I could live with it if we really wanted to try it. I could also be wrong or change my mind later. But *right now*, it feels like a small step in the wrong direction.

My general feeling is that per-file conflict simplification is quite important and safer, while the "deduplication" is much less important and does not feel safe to me. Some of the things I worry about are:

1. Cases where a tree is conflicted but each file in it is not conflicted, creating unresolvable conflicts like https://github.com/martinvonz/jj/issues/3223. In fact, this is a concern even when simplifying conflicts per-file without deduplication. `jj` must already be doing something to avoid these when creating a merge commit of 2 commits that edit disjoint sets of files. 
	
2. Cases where editing a conflicted commit and rebasing descendants results in weird conflict behavior in the descendants, e.g. the conflicts becoming more complicated when they should be becoming simpler.

3. Having too many ad-hoc conflict simplification rules that make corner cases difficult to reason about. The rule "if we have X as base and X as a side, X-X cancels out" fits very well with the patch theory of conflicts; de-duplicating pairs does not AFAICT.

One thing I do *not* think is very important is the use-case of editing conflicts with more than 2 sides in a text editor. To me, this seems to work only in very basic cases and only for very specific users. I'd rather `jj` encouraged people more strongly to resolve ancestor commits with simpler conflicts and/or provided better ways to decompose conflicts.

> do you prefer a separate PR or just separate commits at the end of this PR? If it's a separate PR, what's the reason for that?

I prefer a separate PR, since I'm not sure about the idea, but it's easy to change our minds about whether or not it should be a separate PR. Splitting the commits would already be very helpful; I'd like to see is how much incremental complexity "deduplication" adds over per-file simplification and, less importantly, which tests it affects (especially if there are some that correspond to natural situations).

Reply:
Circumstances led us to just do simplification in this PR, I believe.